hooks/bootstrap.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) 2018 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import contextlib
+
+from sgtk import get_hook_baseclass
+
+
+class BootstrapHook(get_hook_baseclass()):
+    """
+    The bootstrap hook can be added to a configuration's core hooks folder
+    (``config/core/hooks``) in order to override some of the functionality of
+    the ToolkitManager's bootstrapping process.
+
+    The hook will be instantiated only after a configuration has been selected
+    by the ``ToolkitManager``. Therefore, the ``download_bundle`` method
+    will not be invoked to download a configuration. However, the Toolkit Core,
+    applications, frameworks and engines will be downloaded through the hook.
+    """
+
+    def init(self, shotgun, pipeline_configuration_id, configuration_descriptor, **kwargs):
+        """
+        This method is invoked when the hook is instantiated.
+
+        It can be used to cache information so it can be reused, for example a
+        shotgun connection to external site.
+
+        The ``kwargs`` is there for forward compatibility with future versions of Toolkit.
+
+        :param shotgun: Connection to the Shotgun site.
+        :type shotgun: :class:`shotgun_api3.Shotgun`
+
+        :param pipeline_configuration_id: Id of the pipeline configuration we're bootstrapping into.
+            If None, the ToolkitManager is bootstrapping into the base configuration.
+        :type pipeline_configuration_id: int
+
+        :param configuration_descriptor: Configuration the manager is bootstrapping into.
+        :type configuration_descriptor: :class:`sgtk.descriptor.ConfigDescriptor`
+        """
+        self.shotgun = shotgun
+        self.pipeline_configuration_id = pipeline_configuration_id
+        self.configuration_descriptor = configuration_descriptor
+
+    def download_bundle(self, descriptor, **kwargs):
+        """
+        This method is invoked during bootstrapping when downloading the core or
+        the bundles used by a configuration.
+
+        Your override must download the descriptor's content before the method
+        ends or the bundle cache will be left in an inconsistent state.
+
+        The safest way to download a bundle into the bundle cache is by using the
+        :meth:`BootstrapHook._open_write_location` method with the `with` statement.
+
+        .. example::
```

Comment:
do we need `_open_write_location`?

Also here, i think we should provide a commented out example of what this looks like with a 'local app store cache' - so that a studio wanting to use this pattern could pretty much just uncomment the code, set the custom entity id for the custom entity they want to use, and be up and running with this hook in a matter of minutes.

Could we just say:

```
.. example::
with descriptor._io_descriptor.open_write_location() as write_location:
    # open a write transaction to download the contents from a remote location into our bundle cache
    --- here we provide the actual code to cache from the shotgun site ---
    
# once we exit the with statement, all files pushed into the write_location will be
# moved by toolkit into their final location in the bundle cache safely. This 
# transaction helps ensure that the bundle cache doesn't include any half
# processed or corrupt items in the case when multiple users are concurrently
# downloading or when an a download is interrupted half way.
```

Reply:
Yes, we need the method because we do not want to expose io_descriptor to API users.

Good point about adding a commented out example.